os/android/iahwc2.cpp

```diff
@@ -230,19 +230,34 @@ static inline void supported(char const *func) {
 HWC2::Error IAHWC2::CreateVirtualDisplay(uint32_t width, uint32_t height,
                                          int32_t *format,
                                          hwc2_display_t *display) {
-  *display = (hwc2_display_t)(virtual_display_index_ + HWC_DISPLAY_VIRTUAL +
-                              VDS_OFFSET);
+  spin_lock_.lock();
+  uint32_t free_display_index = 0;
+  for (uint32_t disp_index = 0; disp_index < virtual_display_index_;
+       disp_index++) {
+    auto it = virtual_displays_.find(disp_index);
+    if (it == virtual_displays_.end()) {
+      free_display_index = disp_index;
+      break;
+    }
+  }
+  if (!free_display_index) {
+    free_display_index = virtual_display_index_;
+    virtual_display_index_++;
```

Comment:
I think virtual_display_index_ is meant to keep track of virtual display count created or destroyed by the corresponding helpers :)    Incrementing it here without actually having a virtual display created may result to instability.  See CreateVirtualDisplay() and DestroyVirtualDisplay() where this index is being maintained.

Reply:
This is intended, actually decreasing index at virtual display destroy is the root cause of the live virtual display instance overwritten. This leads into HWC crash. The reason of it is VDS creation & deletion might be interleaved. We applied reusing freed display index slots approach that ensures ACRN guests works as expected since it hard coded just using first few display index slots.

Reply:
I see what you're saying and this sounds like it is effective in resolving the crash.  However, what happens if the newly incremented index is referenced while it does not yet have a virtual display attached to it?  Is this a possible scenario?

Reply:
Yes, this is possible. Per evaluation, this scenario is not harmful :-)

Reply:
Okay, keeping fingers crossed :)  It's still a bit risky using a shared flag (virtual_display_index_) for this purpose since there is always a possibility of a race condition between CreateVirtualDisplay and DestroyVirtualDisplay.  When (free_display_index == 0), could you call IAHWC2::CreateVirtualDisplay() so that it will go thru the normal channel of incrementing virtual_display_index_?

Reply:
Don't worry :-) HWC's DestroyVirtualDisplay method will just return if the display is invalid. The underlying caller of CreateVirtualDisplay and DestroyVirtualDisplay is SF. SF only remove the VDS created by itself.